File: mvp-rxlifecycle/src/demo/java/com/ayvytr/mvprxlifecycle/mvp/presenter/MainPresenter.java
package com.ayvytr.mvprxlifecycle.mvp.presenter;

import com.ayvytr.logger.L;
import com.ayvytr.mvpbase.BasePresenter;
import com.ayvytr.mvprxlifecycle.RxUtils;
import com.ayvytr.mvprxlifecycle.mvp.contract.MainContract;
import com.ayvytr.mvprxlifecycle.mvp.model.MainModel;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

/**
 * @author admin
 */
public class MainPresenter extends BasePresenter<MainContract.Model, MainContract.View> {
    public MainPresenter(MainContract.View rootView) {
        super(new MainModel(), rootView);
    }

    public void getData() {
        mModel.getData()
              .compose(RxUtils.<Long>bindToLifecycle(mView))
              .compose(RxUtils.<Long>subscribeIo(mView))
              .subscribe(new Observer<Long>() {
                  @Override
                  public void onSubscribe(Disposable d) {
                      L.e(Thread.currentThread().getName());
                  }

                  @Override
                  public void onNext(Long aLong) {
                      L.e(Thread.currentThread().getName(), aLong);
                      mView.showInterval(aLong);
                  }

                  @Override
                  public void onError(Throwable e) {
                      L.e(Thread.currentThread().getName());
                  }

                  @Override
                  public void onComplete() {
                      L.e(Thread.currentThread().getName());
                  }
              });
    }

    public void getWeather() {
        mModel.getWeather()
              .compose(RxUtils.<String>bindToLifecycle(mView))
              .compose(RxUtils.<String>subscribeIo(mView))
              .subscribe(new Consumer<String>() {
                  @Override
                  public void accept(String weather) {
                      mView.showWeather(weather);
                  }
              });
    }
}

File: mvptest/src/main/java/com/ayvytr/mvptest/presenter/MainPresenter.kt
package com.ayvytr.mvptest.presenter

import com.ayvytr.mvpbase.BasePresenter
import com.ayvytr.mvprxlifecycle.kotlin.RxUtils
import com.ayvytr.mvprxlifecycle.kotlin.bindToLifecycle
import com.ayvytr.mvptest.contract.MainContract
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

class MainPresenter : BasePresenter<MainContract.Model, MainContract.View> {

    constructor() {}

    constructor(model: MainContract.Model, view: MainContract.View) : super(model, view) {}

    constructor(view: MainContract.View) : super(view) {
    }

    fun timer() {
        Observable.interval(1, TimeUnit.SECONDS)
            .bindToLifecycle(mView)
            .compose(RxUtils.subscribeIo(null))
            .subscribe{
                mView.showTimer(it)
            }
    }
}


